### DISCUSSION TITLE
RFC: Dynamic UI Rendering

### DISCUSSION THREAD

**Discussion by barshimi** [amplication/amplication#4362] [Ideas]
With the entrance of plugins to our build, we need to think about the cases the dashboard look/render will be different according to the selected plugin. for example using relational DB vs non-relational DB. This will completely change the presented data and will change the rendered form. also it will change the API request accordingly. I suggest thinking about using configured schema that will be coming from server and will determine the way the client will look like and which structure the API request will look like. another solution that can be used is server rendering although the last one is a completely client refactor. By using this methodology it will help us render each dashboard according to the specific build.

**Comment by arielweinberger** [comment]
Overall sounds interesting. Can you try and come up with one or two examples/scenarios where plugins will affect the flow? This will help make the suggestion clearer.

**Comment by yuval-hazaz** [comment]
@barshimi to make sure I understand correctly, are you referring only to the impact of how we generate the client? Or is that something that will also impact the server?

**Reply by barshimi** [reply]
impact both

**Reply by yuval-hazaz** [reply]
So I don't think I understand the proposal. 
Can you provide some kind of a user story or a flow from a user perspective?

**Reply by barshimi** [reply]
lets say that plugin for db X and db Y require different settings (form) and different API request and schema validation. my proposal is to attache to settings, in this case, the UI settings which dictate the way the page will be render, the form will be validate and the way CRUD requests will be send. If we want to take it further, we can set the server to be able to accept different API structure according to the plugin or the related settings. this will give us the option to have one API for different settings/plugin apart from the ability to render pages according to the used plugin.

**Reply by yuval-hazaz** [reply]
I think that our goal is to be able to generate an API that can support anything behind it (DB etc) and expose the relevant methods and types to the client. Anyway, we dynamically generate the API so we can provide the client with types to represent the DB 


On the platform side (server/dsg), we should be very dynamic and should be able to support different use cases and settings - but on the generated code, we need to provide the end developer with a very simple API that knows nothing about the dynamic options that are not relevant to their specific service.
